projects/common/vk_renderer.cpp

```diff
@@ -3453,6 +3457,283 @@ HRESULT CompileHLSL(
+    if (SLANG_FAILED(globalSession->createSession(sessionDesc, compileSession.writeRef())))
+    {
+        return COMPILE_ERROR_INTERNAL_COMPILER_ERROR;
+    }
+
+    // Load source
+    slang::IModule* pSlangModule = nullptr;
+    {
+        Slang::ComPtr<slang::IBlob> diagBlob;
+
+        pSlangModule = compileSession->loadModuleFromSourceString("grex-module", "grex-path", shaderSource.c_str(), diagBlob.writeRef());
+        if (pSlangModule == nullptr)
+        {
+            if (pErrorMsg != nullptr)
+            {
+                *pErrorMsg = std::string(static_cast<const char*>(diagBlob->getBufferPointer()), diagBlob->getBufferSize());
+            }
+
+            return COMPILE_ERROR_INTERNAL_COMPILER_ERROR;
+        }
+    }
+
+    Slang::ComPtr<slang::IBlob> spirvCode;
+    if (isTargetLibrary)
+    {
+        //
+        // NOTE: This may not be the most correct way to do it, but it works for now
+        //
+
+        // Create compile request
+        std::unique_ptr<SlangCompileRequest, void (*)(SlangCompileRequest*)> compileRequest(nullptr, nullptr);
+        {
+            SlangCompileRequest* pCompileRequest = nullptr;
+            auto                 slangRes        = compileSession->createCompileRequest(&pCompileRequest);
+            if (SLANG_FAILED(slangRes))
+            {
+                return COMPILE_ERROR_INTERNAL_COMPILER_ERROR;
+            }
+
+            compileRequest = std::unique_ptr<SlangCompileRequest, void (*)(SlangCompileRequest*)>(pCompileRequest, spDestroyCompileRequest);
+        }
+
+        // Add translation unit
+        auto index = compileRequest->addTranslationUnit(SLANG_SOURCE_LANGUAGE_SLANG, nullptr);
+        compileRequest->addTranslationUnitSourceString(index, "grex-path", shaderSource.c_str());
+
+        // Compile
+        auto slangRes = compileRequest->compile();
+        if (SLANG_FAILED(slangRes))
+        {
+            if (pErrorMsg != nullptr)
+            {
+                Slang::ComPtr<slang::IBlob> diagBlob;
+
+                slangRes = compileRequest->getDiagnosticOutputBlob(diagBlob.writeRef());
+                if (SLANG_SUCCEEDED(slangRes))
+                {
+                    *pErrorMsg = std::string(static_cast<const char*>(diagBlob->getBufferPointer()), diagBlob->getBufferSize());
+                }
+                else
+                {
+                    // Something has gone really wrong
+                    assert(false && "failed to get diagnostic output blob");
+                }
+            }
+
+            return COMPILE_ERROR_COMPILE_FAILED;
+        }
+
+        // Get SPIR-V
+        slangRes = compileRequest->getTargetCodeBlob(0, spirvCode.writeRef());
+        if (SLANG_FAILED(slangRes))
+        {
+            if (pErrorMsg != nullptr)
+            {
+                *pErrorMsg = "unable to retrieve SPIR-V blob for library";
+            }
+
+            return COMPILE_ERROR_INTERNAL_COMPILER_ERROR;
```

Comment:
Feels like you might want some kind of SLANG_CALL_SAFE() macro or function to deal with this kind of stuff? Instead of calling it all over the place?